crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -0,0 +1,4433 @@
+                .union([expected].as_slice().into());
+            // Skip leading unexpected tokens
+            self.skip_until(expected_set);
+
+            self.eat(expected);
+        }
+    }
+
+    fn add_error(&mut self, error: ParseErrorType, range: TextRange) {
+        self.errors.push(ParseError {
+            error,
+            location: range,
+        });
+    }
+
+    fn skip_until(&mut self, token_set: TokenSet) {
+        while !self.at_ts(token_set) {
+            self.next_token();
+        }
+    }
+
+    fn at(&mut self, kind: TokenKind) -> bool {
+        self.current_kind() == kind
+    }
+
+    fn at_ts(&mut self, ts: TokenSet) -> bool {
+        ts.contains(self.current_kind())
+    }
+
+    fn at_expr(&mut self) -> bool {
+        self.at_ts(EXPR_SET)
+    }
+
+    fn at_simple_stmt(&mut self) -> bool {
+        self.at_ts(SIMPLE_STMT_SET2)
+    }
+
+    fn at_compound_stmt(&mut self) -> bool {
+        self.at_ts(COMPOUND_STMT_SET)
+    }
+
+    fn src_text(&self, mut range: TextRange) -> &'src str {
+        // This check is to prevent the parser from panicking when using the
+        // `parse_expression_starts_at` function with an offset bigger than zero.
+        //
+        // The parser assumes that the token's range values are smaller than
+        // the source length. But, with an offset bigger than zero, it can
+        // happen that the token's range values are bigger than the source
+        // length, causing the parser to panic when calling this function
+        // with such ranges.
+        //
+        // Therefore, we fix this by creating a new range starting at 0 up to
+        // the source length - 1.
+        //
+        // TODO: Create the proper range here.
+        let src_len = self.source.len();
+        if range.start().to_usize() > src_len || range.end().to_usize() > src_len {
+            range = TextRange::new(
+                0.into(),
+                (self.source.len() - 1)
+                    .try_into()
+                    .expect("source length is bigger than u32 max"),
+            );
+        }
+        &self.source[range]
+    }
+
+    fn current_range(&mut self) -> TextRange {
+        self.lookahead(0).1
+    }
+
+    /// Parses elements enclosed within a delimiter pair, such as parentheses, brackets,
+    /// or braces.
+    ///
+    /// Returns the [`TextRange`] of the parsed enclosed elements.
+    fn parse_delimited(
+        &mut self,
+        allow_trailing_delim: bool,
+        opening: TokenKind,
+        delim: TokenKind,
+        closing: TokenKind,
+        mut func: impl FnMut(&mut Parser<'src, I>),
+    ) -> TextRange {
+        let start_range = self.current_range();
+        assert!(self.eat(opening));
+
+        self.parse_separated(
+            allow_trailing_delim,
+            delim,
+            [closing].as_slice(),
+            |parser| {
+                func(parser);
+                // Doesn't matter what range we return here
```

Comment:
Do you mean the empty range that is returned? 

It's not worth changing `func` to return a `TextRange` because, we don't need the range created by `parse_separated` since we only care about the range of the delimiters, e.g., `()`, `[]` etc.